### ISSUE TITLE
Feature: support config rule comments.ignore-regex

### ISSUE THREAD

**Issue by mikaello** [adrienverge/yamllint#545]
In many of our YAML files, we got markers starting with `# renovate:` and those can be quite long depending on how much information we need to include in this marker, they cannot be divided into multiple lines, thus often violating the `rules.line-length` rule.

I have seen similar problems with `#cloud-info`, violating the `rules.comments.require-starting-space: true` rule.

I, therefore, propose (as has been done by @adrienverge multiple times already (https://github.com/adrienverge/yamllint/issues/428#issuecomment-991512486, https://github.com/adrienverge/yamllint/issues/416#issuecomment-951981783, https://github.com/adrienverge/yamllint/issues/361#issuecomment-784197416)) to add the option, `comments.ignore-regex`, example usage:

```yaml
rules:
  comments:
    ignore-regex:
      - ^#!\s*
      - ^#cloud-config
      - ^# renovate:
 ```

**Comment by lyda** [comment]
Is this going in at some point?  It would be really useful.

**Comment by adrienverge** [comment]
> Is this going in at some point?

Not that I'm aware of. Do you want to implement it?

**Comment by ghostwheel42** [comment]
Hi. I've implemented a firt try at this (no tests yet) here: https://github.com/adrienverge/yamllint/compare/master...ghostwheel42:yamllint:comment_ignore_regex
The "full" regular expression is maybe not really helpful, because the checker does not work line-based but token-based.
I'm adding null bytes when a comment is indented (column > 1), but this seems really ugly.

Is it really necessary to allow a regular expession here?
My use-cases would do fine with just a "ignore-prefix".

What do you think?

**Comment by adrienverge** [comment]
Hello, I still think that using regular expessions is the option that gives the most flexibility (as opposed to options that allow ignoring comments based on prefixes only, which may not satisfy all users). For info, the rule `quoted-strings` already [accepts PCRE regexes in its configuration](https://yamllint.readthedocs.io/en/stable/rules.html#module-yamllint.rules.quoted_strings).

**Comment by ghostwheel42** [comment]
> Hello, I still think that using regular expessions is the option that gives the most flexibility (as opposed to options that allow ignoring comments based on prefixes only, which may not satisfy all users).

I'm not really against using a regex here, but checking a "quoted string" is slightly different than checking a comment. It is not only about matching the content but also the surroundings - in this case: "Does the comment start on the first column?" or even "Is it on the same line?".
This is why I prepended col*"\0" when matching the comment to have ^ match the start of the line and not the comment. I could change it to just match like the "quoted string" test, but then all comments in this code would be valid:

```
#cloud-config
- foo: bar #cloud-config
```

> For info, the rule `quoted-strings` already [accepts PCRE regexes in its configuration](https://yamllint.readthedocs.io/en/stable/rules.html#module-yamllint.rules.quoted_strings).

Sure, this is where I "borrowed" some lines.

**Comment by adrienverge** [comment]
Hello, I get your point :+1: But still, I feel that future users could want to allow some comments that don't start a line (e.g. `- yaml line  #noqa`).

If I understand correctly, your implementation correctly implements `^` and `$` for the whole line (not just the comment).

So this conf:
```yaml
rules:
  comments:
    ignore-regex:
      - ^#cloud-config
      - allow this$
```
would yield:

```yaml
#cloud-config                     ← OK
- item  #cloud-config             ← error
- item  # allow this              ← OK
- item  # allow this.             ← error
```
?

**Comment by ghostwheel42** [comment]
Yes, this is what it will do. But it will not be possible to match the content before the inline comment.
It appears "hacky" to me, but it will do what you described.

I can add some tests and create a pr if you like.
